Security considerations for client_metadata parameters (#735)
openid/OpenID4VP: 1.1/openid-4-verifiable-presentations-1_1.md

In case of using OpenID4VP over DC API, as defined in (#dc_api), it is at the discretion of the Wallet whether it validates the signature on the Request Object following the processing rules defined by a relevant Client Identifier Prefix. Factors that influence the Wallet's decision include, but are not limited to, the trust framework the Wallet supports, the specific policies defined by the Issuers or ecosystem, and profiles of this specification.

> * `redirect_uri`: This prefix value indicates that the original Client Identifier part (without the prefix `redirect_uri:`) is the Verifier's Redirect URI (or Response URI when Response Mode `direct_post` is used). The Verifier MAY omit the `redirect_uri` Authorization Request parameter (or `response_uri` when Response Mode `direct_post` is used). All Verifier metadata parameters MUST be passed using the `client_metadata` parameter defined in (#new_parameters). An example Client Identifier value is `redirect_uri:https://client.example.org/cb`. Requests using the `redirect_uri` Client Identifier Prefix cannot be signed because there is no method for the Wallet to obtain a trusted key for verification. Therefore, implementations requiring signed requests cannot use the `redirect_uri` Client Identifier Prefix.

Reviewer: I suggest we put at least this change into 1.0 errata too.

Author: Done

* `x509_san_dns`: When the Client Identifier Prefix is `x509_san_dns`, the original Client Identifier (the part after the `x509_san_dns:` prefix) MUST be a DNS name and match a `dNSName` Subject Alternative Name (SAN) [@!RFC5280] entry in the leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header [@!RFC7515] of the signed request object. The Wallet MUST validate the signature and the trust chain of the X.509 certificate. The following requirement applies unless the interaction is using the DC API as defined in (#dc_api): If the Wallet can establish trust in the Client Identifier authenticated through the certificate, e.g. because the Client Identifier is contained in a list of trusted Client Identifiers, it may allow the client to freely choose the `redirect_uri` value. If not, the FQDN of the `redirect_uri` value MUST match the Client Identifier without the prefix `x509_san_dns:`. Example Client Identifier: `x509_san_dns:client.example.org`.

> * `x509_hash`: When the Client Identifier Prefix is `x509_hash`, the original Client Identifier (the part without the `x509_hash:` prefix) MUST be a hash and match the hash of the leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header parameter [@!RFC7515] of the signed request object. The value of `x509_hash` is the base64url-encoded value of the SHA-256 hash of the DER-encoded X.509 certificate. The Wallet MUST validate the signature and the trust chain of the X.509 leaf certificate. Example Client Identifier: `x509_hash:Uvo3HtuIxuhC92rShpgqcT3YXwrqRxWEviRiA0OZszk`

Reviewer: Add the missing fullstop (the 1.0 change already added this):

```suggestion
* `x509_hash`: When the Client Identifier Prefix is `x509_hash`, the original Client Identifier (the part without the `x509_hash:` prefix) MUST be a hash and match the hash of the leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header parameter [@!RFC7515] of the signed request object. The value of `x509_hash` is the base64url-encoded value of the SHA-256 hash of the DER-encoded X.509 certificate. The Wallet MUST validate the signature and the trust chain of the X.509 leaf certificate. Example Client Identifier: `x509_hash:Uvo3HtuIxuhC92rShpgqcT3YXwrqRxWEviRiA0OZszk`.
```

MUST perform its own security checks on the returned Credentials and
Presentations.

> ## Source of verifier metadata considerations

Reviewer 2: Nit: match the style of the other headings in this section:
```suggestion
## Source of Verifier Metadata
```

> Wallet implementers need to choose a source of client metadata they can trust within their own ecosystem or implementation. This should determine whether it is safe to use or display to a User. For example, a client name obtained from an X.509 certificate can be trusted, whereas a client name from the `client_metadata` parameter cannot.

Reviewer: Maybe something more like this:

```suggestion
Wallet implementers need to choose a source of client metadata they can trust within their own ecosystem or implementation. This should determine whether it is safe to use or display to a End-User. For example, a client name obtained from an X.509 certificate that was used to a signed request, or a client name obtained from OpenID Federation from a signed requested, can be trusted, whereas a client name from the `client_metadata` parameter cannot.
```

> Wallet implementers need to choose a source of client metadata they can trust within their own ecosystem or implementation. This should determine whether it is safe to use or display to a User. For example, a client name obtained from an X.509 certificate can be trusted, whereas a client name from the `client_metadata` parameter cannot.

Reviewer 2: A competing suggestion to @jogu's:
```suggestion
Wallet implementers need to choose a source of client metadata they can trust within their own ecosystem or implementation. The trust in the client metadata determines whether it is safe to use or display to a User. For example, a client name obtained from an X.509 certificate used in a signed request can be trusted, whereas a client name from the `client_metadata` parameter cannot.
```

Reviewer: Updated version of this - I think the 'can be trusted' in the original sentence is a bit too strong:

```suggestion
Wallet implementers need to choose a source of client metadata they can trust within their own ecosystem or implementation. The trust in the client metadata determines whether it is safe to use or display to a User. For example, a client name obtained from an X.509 certificate used in a signed request can potentially be trusted, whereas a client name from the `client_metadata` parameter cannot.
```

I think it'd be good to mention OpenID Federation as an example too, but that's not a hill I feel I need to die on.

Reviewer 3: ```suggestion
Wallet implementers need to choose a sources of client metadata they can trust within their own ecosystem or implementation. This should determine whether it is safe to use or display to a User. For example, a client name obtained from an X.509 leaf certificate that chains to a trust anchor the Wallet trusts may be appropriate to be displayed to an End-User, whereas a client name from the `client_metadata` parameter is not suitable to display. 
```

Reviewer 3: I've added my proposal to that list. +sources. Improved example - X.509 solo looked quite generic. I support the addition of OpenID Federation, but that's just an example, so that single one is enough.

> It is recommended to use`client_metadata` parameters only for client-supplied parameters that can be used untrusted, for example, a public key. 

Reviewer 2: Nit:
```suggestion
It is recommended to use `client_metadata` parameters only for client-supplied parameters that can be used untrusted, for example, a public key. 
```

> It is recommended to use `client_metadata` parameters only for client-supplied parameters that can be used untrusted, for example, a public key. 

Reviewer: Probably need to clarify this - a public key used for signature verification probably can't be trusted:

```suggestion
It is recommended to use `client_metadata` parameters only for client-supplied parameters that can be used untrusted, for example, a public key used to encrypt the response. 
```

> Ecosystems can use other ecosystem-specific trusted data sources as long as they conduct a proper trust evaluation. 

Reviewer: I think we have to be careful not to conflict with this text:

<img width="784" height="487" alt="Image" src="https://github.com/user-attachments/assets/ff2d98f6-d1a7-4df5-84c8-f12dab7d05fb" />

"It is recommended" probably needs to be reworded, and 'Ecosystems' we used 'profiles' in the other text.

Reviewer 2: @jogu What specifically is the problem with the "it is recommended" text? Just that it reads as a normative statement?

And what about something like this:
```suggestion
Ecosystems can define other ecosystem-specific trusted data sources in a profile as long as they conduct a proper trust evaluation.
```

Reviewer 4: @jogu where do you think it conflicts? Is it the example of the public key and should we be more specific that it is about specific, ephemeral public keys as described in the section?

In the spec I also often notice we use "ecosystems or profiles" instead of the one or the other. Maybe that applies here as well.

Reviewer 2: @jogu Would be great to have some more detail on this.

Reviewer: I think my concern was in the conflict between

> It is recommended (for Wallet implementers) to use `client_metadata` parameters...

and

> Other metadata parameters MUST be ignored unless a profile <...>'

If the recommendation is scoped to profiles that's better. But maybe something like:

> Profiles MAY define additional `client_metadata` parameters for client-supplied parameters that can be used untrusted, for example, a public key.

(Pretty sure that can be improved upon)

MUST perform its own security checks on the returned Credentials and
Presentations.

## Source of Verifier Metadata

> Wallet implementers need to choose a sources of client metadata they can trust within their own ecosystem or implementation. This should determine whether it is safe to use or display to a User. For example, a client name obtained from an X.509 leaf certificate that chains to a trust anchor the Wallet trusts may be appropriate to be displayed to an End-User, whereas a client name from the `client_metadata` parameter is not suitable to display. 

Reviewer: Repeating [my above suggestion](https://github.com/openid/OpenID4VP/pull/735/changes#r3748493731) as it needs applied here too:

```suggestion
Wallet implementers need to choose a source of Verifier metadata they can trust within their own ecosystem or implementation, including determining whether metadata is safe to use or display to an End-User. For example, a Verifier name obtained from an X.509 leaf certificate that chains to a trust anchor the Wallet trusts may be suitable to display to an End-User, whereas a Verifier name from the `client_metadata` parameter is not suitable to display.
```

> It is recommended to use `client_metadata` parameters only for client-supplied parameters that can be used untrusted, for example, a public key used to encrypt the response. 

Reviewer: Repeating [my above suggestion](https://github.com/openid/OpenID4VP/pull/735/changes#r3720296324) as it needs applied here too:

```suggestion
Profiles of this specification MAY define additional `client_metadata` parameters, as described in (#new_parameters). It is RECOMMENDED that such parameters are limited to client-supplied values that can safely be used without being trusted. One example of such a value is a public key used to encrypt the Authorization Response.
```

> Ecosystems can use other ecosystem-specific trusted data sources as long as they conduct a proper trust evaluation. 

Reviewer: Repeating 1.0 suggestion from above:

```suggestion
Wallets MAY use other ecosystem-specific trusted sources of Verifier metadata, provided the Wallet is able to establish trust in those sources. As described in (#new_parameters), authoritative data obtained from such sources takes precedence over the values passed in the `client_metadata` parameter.
```
